### PR TITLE
Lg/rust js type mapping

### LINKED ISSUE #164: User’s guide: Information for rust developers
The new user’s guide, added in #158, does not fully address the Rust-specific workflow and type mappings.

## Story

As a rust canister developer, I want Rust-specific Candid documentation, so that I can build Rust canisters that use Candid.

## Demo

* Go to https://github.com/dfinity/candid/blob/master/docs/modules/candid-guide/index.adoc
* Section “How do I use Candid as Rust canister developer?” has useful information
* In the reference section, all types describe their mapping to rust types.

### PR DESCRIPTION
**Pull request by lsgunnlsgunn**
* Fix #164 #165 
* Add `How to` section for each language

### CODE REVIEW

**Comment by chenyan-dfinity** [comment]
@lsgunnlsgunn I rewrite the `how to` section, please take a look.

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 1]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+  public func greet() : async Text {
+    let result = await Counter.inc(1);
+    "The current counter is " # Nat.toText(result)
+  };
+}
+....
 
-If you are writing a service from scratch in Motoko, the Motoko compiler automatically translates the signature of your canister’s top-level actor or actor class into a Candid description, and the `+dfx build+` command ensures that the service description is properly referenced where it needs to be.
+The import dependency of counter canister is handled by `+dfx build+`, which ensures that the canister id and the Candid
+description is correctly feed into the {proglang} compiler. The {proglang} compiler then translate the Candid type into
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:99]:
```suggestion
The {proglang} compiler then translates the Candid type into the appropriate native Motoko type. This translation enables you to call the `inc` method natively—as if it were a Motoko function—even if the `counter` canister is implemented in a different language and even if you do not have the source code for the imported canister.
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 2]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+native Motoko type and enables you to call the `inc` method natively, even if the counter canister is implemented in a
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:100]:
```suggestion
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 3]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+different language and you do not have the source code. For additional information on the type mapping
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:101]:
```suggestion
For additional information on the type mapping
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 4]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+between Candid and {proglang}, you can consult the link:candid-types{outfilesuffix}[Supported types] reference section.
 
-For example, if you have a multi-canister project written in {proglang}, `+dfx build+` ensure that any clients of your Motoko service import the generated Candid description of that service.
+The {proglang} compiler and `+dfx build+` also auto-generate the Candid description for the `hello` canister to allow
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:104]:
```suggestion
The {proglang} compiler and `+dfx build+` command also auto-generate the Candid description for the `hello` canister to allow
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 5]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+
+#[update]
+async fn greet() -> String {
+    let result = Counter::inc(1.into()).await;
+    format!("The current counter is {}", result)
 }
 ....
 
-You could configure the project `+dfx.json+` file to set the path to the service description like this:
+The import dependency of counter canister is handled by `+dfx build+`, which ensures that the canister id and the Candid
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:128]: Similar suggestion as above 
```suggestion
When the import macro on the `counter` canister— the `#[import(canister = "counter")]` declaration—is processed by the `+dfx build+` command, the `+dfx build+` command ensures that the `counter` canister identifier and the Candid description are passed to the Rust CDK correctly.
The Rust CDK then translates the Candid type into the appropriate native Rust type. 
This translation enables you to call the `inc` method natively—as if it were a Rust function—even if the `counter` canister is implemented in a different language and even if you do not have the source code for the imported canister.
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 6]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
-For examples of how to write Candid service descriptions for Rust programs, see the link:https://github.com/dfinity/cdk-rs/tree/next/examples[Rust CDK examples] and the link:../rust-guide/rust-intro{outfilesuffix}[Rust tutorials].
+There is also an experimental feature to generate Candid service description automatically, see this https://github.com/dfinity/candid/blob/master/rust/candid/tests/types.rs#L99[test case] as an example.
+
+For additional information and libraries to help you create Candid services or canisters in Rust, see the documentation for the https://docs.rs/candid/[Candid crate], link:https://github.com/dfinity/cdk-rs/tree/next/examples[Rust CDK examples] and the link:../rust-guide/rust-intro{outfilesuffix}[Rust tutorials].
+
+== Interact with a service from JavaScript
 
-For additional information and libraries to help you create Candid services or clients in Rust, you can download and use the https://crates.io/crates/candid[Candid crate].
+The https://www.npmjs.com/package/@dfinity/agent[dfinity/agent npm package] includes support for
+importing canisters via Candid.
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:150]:
```suggestion
importing canisters using Candid.
```

File: docs/modules/candid-guide/pages/candid-howto.adoc [hunk 7]
```diff
@@ -50,46 +77,95 @@ For example, if a function takes an array, you might need to specify the number
...
+import counter from 'ic:canisters/counter';
+import BigNumber from 'bignumber.js';
+async () => {
+  const result = await counter.inc(new BigNumber(42));
+  console.log("The current counter is " + result.toString());
+}
+....
 
-The https://github.com/dfinity/agent-js/[Internet Computer agent library for JavaScript] includes support for Candid.
+The import dependency of counter canister is handled by `+dfx build+` and `webpack`, which ensures that the canister id and the Candid description is correctly feed into the JavaScript program. Behind the scenes, the Candid service description is
```
Comment (lsgunnlsgunn) [inline docs/modules/candid-guide/pages/candid-howto.adoc:164]: similar suggestion
```suggestion
When the import dependency of counter canister is processed by the `+dfx build+` command and the `webpack`configuration, this processing ensures that the canister id and the Candid description are passed to the JavaScript program correctly. Behind the scenes, the Candid service description is
```

**Review by lsgunnlsgunn — COMMENTED** [review]
Thanks for your help with this Yan. Much more information than before!

**Pull request merged**